pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -981,6 +890,96 @@ func (c *Controller) updateEDS(ep *v1.Endpoints, event model.Event) {
+	svcName := slice.Labels[discoveryv1alpha1.LabelServiceName]
+	hostname := kube.ServiceHostname(svcName, slice.Namespace, c.domainSuffix)
+	mixerEnabled := c.Env != nil && c.Env.Mesh != nil && (c.Env.Mesh.MixerCheckServer != "" || c.Env.Mesh.MixerReportServer != "")
+
+	endpoints := make([]*model.IstioEndpoint, 0)
+	if event != model.EventDelete {
+		for _, e := range slice.Endpoints {
+			if e.Conditions.Ready != nil && !*e.Conditions.Ready {
+				// Ignore not ready endpoints
+				continue
+			}
+			for _, a := range e.Addresses {
+				pod := c.pods.getPodByIP(a)
+				if pod == nil {
+					// This can not happen in usual case
+					if e.TargetRef != nil && e.TargetRef.Kind == "Pod" {
+						log.Warnf("Endpoint without pod %s %s.%s", a, svcName, slice.Namespace)
+
+						if c.Env != nil {
+							c.Env.PushContext.Add(model.EndpointNoPod, string(hostname), nil, a)
+						}
+						// TODO: keep them in a list, and check when pod events happen !
+						continue
+					}
+					// For service without selector, maybe there are no related pods
+				}
+
+				var labels map[string]string
+				locality, sa, uid := "", "", ""
+				if pod != nil {
+					locality = c.GetPodLocality(pod)
+					sa = kube.SecureNamingSAN(pod)
+					if mixerEnabled {
+						uid = fmt.Sprintf("kubernetes://%s.%s", pod.Name, pod.Namespace)
+					}
+					labels = configKube.ConvertLabels(pod.ObjectMeta)
+				}
+
+				// EDS and ServiceEntry use name for service port - ADS will need to
+				// map to numbers.
+				for _, port := range slice.Ports {
+					var portNum uint32
+					if port.Port != nil {
+						portNum = uint32(*port.Port)
+					}
+					var portName string
+					if port.Name != nil {
+						portName = *port.Name
+					}
+					endpoints = append(endpoints, &model.IstioEndpoint{
+						Address:         a,
+						EndpointPort:    portNum,
+						ServicePortName: portName,
+						Labels:          labels,
+						UID:             uid,
+						ServiceAccount:  sa,
+						Network:         c.endpointNetwork(a),
+						Locality:        locality,
+						Attributes:      model.ServiceAttributes{Name: svcName, Namespace: slice.Namespace},
+					})
+				}
+			}
+		}
+	}
+
+	if log.InfoEnabled() {
+		var addresses []string
+		for _, ss := range slice.Endpoints {
+			for _, a := range ss.Addresses {
+				addresses = append(addresses, a)
+			}
+		}
+		log.Infof("Handle EDS endpoint %s in namespace %s -> %v", svcName, slice.Namespace, addresses)
+	}
+
+	if features.EnableHeadlessService.Get() {
+		if obj, _, _ := c.services.informer.GetIndexer().GetByKey(kube.KeyFunc(svcName, slice.Namespace)); obj != nil {
+			svc := obj.(*v1.Service)
+			// if the service is headless service, trigger a full push.
+			if svc.Spec.ClusterIP == v1.ClusterIPNone {
+				c.XDSUpdater.ConfigUpdate(&model.PushRequest{Full: true, NamespacesUpdated: map[string]struct{}{slice.Namespace: {}}})
+				return
+			}
+		}
+	}
+
+	_ = c.XDSUpdater.EDSUpdate(c.ClusterID, string(hostname), slice.Namespace, endpoints)
```

Comment:
This is not right, there can be many endpointslices for a service. And here you only update one.  Need to think incremental update.

Reply:
hmm yeah. Thats going to make this a lot less efficient since we need to query for all the ep slices every time we get an update for one. Maybe need to be a bit smarter here to make that more efficient